File: tags/release-0.9/src/tileopt.cpp
// File:     tileopt.cpp
// Location: src
// Purpose:  implement TileOpt class
// Author:   Stephen Gold sgold@sonic.net
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the 
Free Software Foundation, either version 3 of the License, or (at your 
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "string.hpp"
#include "tileopt.hpp"


// lifecycle

TileOpt::TileOpt(void) {
    mHasBonus = false;
}

// new TileOpt from String
TileOpt::TileOpt(String const& rString) {
    String copy = rString;

    mHasBonus = false;
    if (copy.Last() == BONUS_CHARACTER) {
        mHasBonus = true;
        copy.Shorten(1);
    }
    mCombo = Combo(copy);
}

// The compiler-generated copy constructor is fine.
// The compiler-generated destructor is fine.


// operators

// The compiler-generated assignment method is fine.

bool TileOpt::operator==(TileOpt const& rOther) const {
    bool const result = (mCombo == rOther.mCombo
        && mHasBonus == rOther.mHasBonus);

    return result;
}

TileOpt::operator Combo(void) const {
    return mCombo;
}

TileOpt::operator String(void) const {
    String result = String(mCombo);
    if (mHasBonus) {
        result += String(BONUS_CHARACTER);
    }

    return result;
}


// misc methods

AttrType TileOpt::Attribute(AttrIndexType index) const {
    AttrType const result = mCombo.Attribute(index);

    return result;
}

AttrIndexType TileOpt::CommonAttribute(TileOpt const& rOther) const {
    AttrIndexType const result = mCombo.CommonAttribute(rOther.mCombo);

    return result;
}

AttrCntType TileOpt::CountMatchingAttributes(TileOpt const& rOther) const {
    AttrCntType const result = mCombo.CountMatchingAttributes(rOther.mCombo);

    return result;
}

String TileOpt::Description(void) const {
    String result = mCombo.Description();
    if (mHasBonus) {
        result += String('+');
    }

    return result;
}

/* static */ TileOpt TileOpt::FromDescription(String const& rDescription) {
    String copy = rDescription;
    copy.Purge();

    TileOpt result;
    if (copy.Last() == '+') {
        result.mHasBonus = true;
        copy.Shorten(1);
    }
    result.mCombo = Combo::FromDescription(copy);

    return result;
}

void TileOpt::SetAttribute(AttrIndexType index, AttrType value) {
    mCombo.SetAttribute(index, value);
}

void TileOpt::SetBonus(bool newBonus) {
    mHasBonus = newBonus;
}


// inquiry methods

bool TileOpt::HasAttribute(AttrIndexType index, AttrType value) const {
    bool const result = mCombo.HasAttribute(index, value);

    return result;
}

bool TileOpt::HasBonus(void) const {
    return mHasBonus;
}

bool TileOpt::IsCompatibleWith(TileOpt const& rOther) const {
    bool const result = mCombo.IsCompatibleWith(rOther);

    return result;
}

bool TileOpt::MatchesDescription(String const& rMatch) const {
    String description = Description();

    // Purge both strings, so that only graphic characters are compared.
    description = description.Purge();
    String const match = rMatch.Purge();

    bool const result = (description == match);

    return result;
}

File: Java/GoldTile/src/goldtile/TileOpt.java
// File:     TileOpt.java
// Location: Java/GoldTile/src/goldtile
// Purpose:  TileOpt class for the Gold Tile Game
/**
 * @author Stephen Gold
 */
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the 
Free Software Foundation, either version 3 of the License, or (at your 
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

package goldtile;
public class TileOpt {
    public Combo combo;
    public boolean hasBonus;
    
    // static fields
    
    private final static char BONUS_CHARACTER = '+';
    
    // constructors
    
    public TileOpt() {
        combo = new Combo();
        hasBonus = false;
    }

    public TileOpt(Combo c) {
        combo = new Combo(c);
    }
    
    /**
     * @param text string-form of a TileOpt to construct
     */
    public TileOpt(String text) {
        String copy;
        if (StringExt.last(text) == BONUS_CHARACTER) {
            hasBonus = true;
            copy = StringExt.shorten(text, 1);
        } else {
            hasBonus = false;
            copy = text;
        }
        combo = new Combo(copy);
    }
    
    /**
     * @param other the TileOpt to be replicated
     */
    public TileOpt(TileOpt other) {
        hasBonus = other.hasBonus;
        combo = new Combo(other.combo);
    }

    // methods
    
    public String description() {
        String result = combo.description();
        
        if (hasBonus)
            result += BONUS_CHARACTER;
        
        return result;
    }
    
    /**
     * @param other the TileOpt to be compared
     */
    public boolean equals(TileOpt other) {
        return hasBonus == other.hasBonus
                && combo.equals(other.combo);
    }
     
    /**
     * @param text description of a TileOpt to construct
     */
    static public TileOpt fromDescription(String text) {
        TileOpt result = new TileOpt();
        
        String copy = text;
        if (StringExt.last(text) == BONUS_CHARACTER) {
            result.hasBonus = true;
            copy = StringExt.shorten(text, 1);
        }
        result.combo = Combo.fromDescription(copy);
        
        return result;
    }
    
    /**
     * @param text the description to be compared
     */
    public boolean matchesDescription(String text) {
        String description = description();

        // Purge both strings, so that only graphic characters are compared.
        description = StringExt.purge(description);
        final String match = StringExt.purge(text);

        return description.equals(match);
    }
    
    public String toString() {
        String result = combo.toString();
        
        if (hasBonus)
            result += BONUS_CHARACTER;
        
        return result;
    }
}
